src/ctelnet.cpp

```diff
@@ -408,40 +424,67 @@ bool cTelnet::sendData(QString& data)
             outData = outgoingDataEncoder->fromUnicode(data).constData();
         } else {
             // Plain, raw ASCII, we hope!
-            // TODO: Moan if the user is trying to send non-ASCII characters out!
+            for (int i = 0, total = data.size(); i < total; ++i) {
+                if (data.at(i).row() || data.at(i).cell() > 127){
+                    QString errorMsg = tr("[ WARN ] - Invalid characters in outgoing data, one or more characters cannot\n"
+                                          "be encoded into the range that is acceptable for the character\n"
+                                          "encoding that is currently set {\"ASCII\"} for the MUD Server."
+                                          "It may not understand what is sent to it.");
+                    postMessage(errorMsg);
+                    break;
+                }
+            }
+            // Even if there are bad characters - try to send it anyway...
             outData = data.toStdString();
         }
 
         if (!mpHost->mUSE_UNIX_EOL) {
-            outData.append("\r\n");
-        } else {
-            outData += "\n";
+            outData += "\r";
         }
+        outData += "\n";
+
+        // outData is using the selected Mud Server encoding here:
+        // we need to cook any byte values from the encoding process that are
+        // 0xff (assuming that there are no Telnet protocol sequences in here):
+        outData = mudlet::replaceString(outData, "\xff", "\xff\xff");
         return socketOutRaw(outData);
     } else {
+
         mpHost->mAllowToSendCommand = true;
         return false;
     }
 }
 
-
+// Data is *expected* to be in the required MUD Server encoding on entry,
+// of course plain ASCII *is* valid for all encodings including Big-5 and GBK,
+// as we do NOT handle the weirdly different EBDIC!!!
 bool cTelnet::socketOutRaw(string& data)
 {
-    if (!socket.isWritable()) {
+    // We were using socket.iswritable() but it was not clear that that was a
```

Comment:
This should be in the Git commit, not as a code comment.

Reply:
Well it also will stop someone reverting (or merge conflict resolving) to going back to using `isWritable()` without due consideration...

Reply:
OK.